backend/tests/test_sso.py

```diff
@@ -0,0 +1,206 @@
+#
+# Copyright (c) 2017-2018, Magenta ApS
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+import base64
+import os
+from urllib.parse import urlparse, parse_qs, urlencode
+
+import settings
+from flask_testing import TestCase
+from mock import patch, MagicMock
+
+from mora.app import app
+from onelogin.saml2.utils import OneLogin_Saml2_Utils as saml_utils
+
+TESTS_DIR = os.path.dirname(__file__)
+
+
+@patch('onelogin.saml2.response.OneLogin_Saml2_Response.is_valid',
+       lambda *x, **y: True)
+class TestSSO(TestCase):
+
+    def create_app(self):
+        app.config['TESTING'] = True
+        app.config['SERVER_NAME'] = "127.0.0.1:5000"
+
+        app.config['AUTH'] = True
+        app.config['SAML_IDP_METADATA_FILE'] = TESTS_DIR + '/sso/idp.xml'
+        return app
+
+    def get_sso_response(self):
+        with open(TESTS_DIR + '/sso/sso_response.xml', 'rb') as sso_response:
+            return base64.b64encode(sso_response.read())
+
+    def get_slo_response(self):
+        with open(TESTS_DIR + '/sso/slo_response.xml', 'rb') as slo_response:
+            return saml_utils.deflate_and_base64_encode(slo_response.read())
+
+    def test_sso_redirects_to_login_with_next(self):
+        r = self.client.get('/saml/sso/?next=http://redirect.me/to/here')
+        url = urlparse(r.location)
+        query = parse_qs(url.query)
+
+        self.assertEqual(302, r.status_code)
+        self.assertEqual('https', url.scheme)
+        self.assertEqual('192.168.1.212', url.netloc)
+        self.assertEqual('/simplesaml/saml2/idp/SSOService.php', url.path)
+        self.assertEqual('http://redirect.me/to/here',
+                         query.get('RelayState')[0])
+
+    def test_sso_redirects_to_login_with_no_next(self):
+        r = self.client.get('/saml/sso/')
+        url = urlparse(r.location)
+        query = parse_qs(url.query)
+
+        self.assertEqual(302, r.status_code)
+        self.assertEqual('https', url.scheme)
+        self.assertEqual('192.168.1.212', url.netloc)
+        self.assertEqual('/simplesaml/saml2/idp/SSOService.php', url.path)
+        self.assertEqual('http://127.0.0.1:5000/',
+                         query.get('RelayState')[0])
+
+    @patch('onelogin.saml2.response.OneLogin_Saml2_Response.is_valid',
+           lambda *x, **y: True)
+    def test_acs_redirects_correctly_with_relaystate(self):
+        data = {
+            'SAMLResponse': self.get_sso_response(),
+            'RelayState': 'http://redirect.me/to/here'
+        }
+        r = self.client.post('/saml/acs/', data=data)
```

Comment:
AFAICT this test doesn't freeze the time. We should do so, and verify that the authentication fails when time is frozen at when the token either isn't valid yet or expired.

Reply:
Good point, I'll do that.

Reply:
The general scenarios I can think of are:

1) Before the assertion is valid.
2) After the assertion expires.
3) When the assertion is signed by another entity/the signature is invalid.
4) When the assertion isn't issued to us.
5) Our behaviour on errors.